rpki-validator/src/main/java/net/ripe/rpki/validator3/domain/metrics/TrustAnchorMetricsService.java

```diff
@@ -0,0 +1,143 @@
+package net.ripe.rpki.validator3.domain.metrics;
+
+
+import io.micrometer.core.instrument.*;
+import lombok.Setter;
+import lombok.extern.slf4j.Slf4j;
+import net.ripe.rpki.validator3.storage.Storage;
+import net.ripe.rpki.validator3.storage.data.Key;
+import net.ripe.rpki.validator3.storage.data.TrustAnchor;
+import net.ripe.rpki.validator3.storage.data.validation.ValidationCheck;
+import net.ripe.rpki.validator3.storage.data.validation.CertificateTreeValidationRun;
+import net.ripe.rpki.validator3.storage.stores.ValidationRuns;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+
+@Slf4j
+@Service
+@Setter
+public class TrustAnchorMetricsService {
+    @Autowired
+    private MeterRegistry registry;
+
+    @Autowired
+    private Storage storage;
+
+    @Autowired
+    private ValidationRuns validationRuns;
+
+    /**
+     * Creating multiple counters for the same trust anchor has no negative side-effects.
+     */
+    private ConcurrentHashMap<String, CertificateTreeValidationMetrics> cetrificateTreeValidationMetrics = new ConcurrentHashMap<>();
+
+    public void update(TrustAnchor ta, CertificateTreeValidationRun vr, long durationMs) {
+        final String uri = ta.getLocations().size() > 0 ? ta.getLocations().get(0) : null;
+        if (uri == null) {
+            log.error("Trust anchor {} does not have a location.", ta.getSubjectPublicKeyInfo());
+            return;
+        }
+
+        if (cetrificateTreeValidationMetrics.containsKey(uri)) {
```

Comment:
Strictly speaking, ConcurrentHashMap doesn't protect from race conditions between calls of containsKey and get in this case. 

This should work atomically 
cetrificateTreeValidationMetrics
   .putIfAbsent(uri, new CertificateTreeValidationMetrics(ta, vr, durationMs))
   .update(vr, durationMs);

Reply:
Or usual map and synchronised block.

Reply:
I was doing the initial update from the constructor. I made this slightly cleaner.

Reply:
It did not matter if the object was constructed twice. I cleaned this by moving the initial update from the constructor and doing that explicitly.